tools/gcs-bucket-mover/gcs_bucket_mover/bucket_details.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2018 Google LLC. All rights reserved. Licensed under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations under
+# the License.
+#
+# Any software provided by Google hereunder is distributed "AS IS", WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, and is not intended for production use.
+"""This is used to track the settings of the source bucket.
+
+It is required so that after the source bucket is deleted, we can create the new one with all of
+the same details and settings.
+"""
+
+
+class BucketDetails(object):
+    """Holds the details and settings of a bucket."""
+
+    # pylint: disable=attribute-defined-outside-init
+    # This is done intentionally so that properties set either in the init or modified in outside
+    # code will be forced to follow the skip rules specified on the command line.
+
+    # pylint: disable=too-many-instance-attributes
+    # All of these attributes relate to the bucket being copied.
+
+    def __init__(self, conf, source_bucket):
+        """Init the class from a source bucket.
+
+        Args:
+            conf: the argparser parsing of command line options
+            source_bucket: a google.cloud.storage.Bucket object that the bucket details should be
+                copied from.
+        """
+
+        # Unless these values are specified on the command line, use the values from the source
+        # bucket
+        self._location = conf.location or source_bucket.location
+        self._storage_class = conf.storageClass or source_bucket.storage_class
+
+        self._set_skip_settings(conf)
+
+        # These properties can be skipped with cmd line params, so use the property setters to
+        # make the checks
+        self.iam_policy = source_bucket.get_iam_policy()
```

Comment:
This is done intentionally because command line flags can specify if a specific setting should be skipped. The check for this is made inside the iam_policy setter. In order to avoid duplicating the check inside __init__, I set iam_policy on this line so it will use the property setter instead of the private variable (along with the other values that have checks on the command line flags).